archive/100-theorems-list/81_sum_of_prime_reciprocals_diverges.lean

```diff
@@ -0,0 +1,281 @@
+/-
+Copyright (c) 2021 Manuel Candales. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Manuel Candales
+-/
+import topology.instances.ennreal
+import algebra.squarefree
+
+/-!
+# Divergence of the Prime Reciprocal Series
+This file proves Theorem 81 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/).
+The theorem states that the sum of the reciprocals of all prime numbers diverges.
+The formalization follows Erdős's proof by upper and lower estimates.
+
+## References
+https://en.wikipedia.org/wiki/Divergence_of_the_sum_of_the_reciprocals_of_the_primes
+-/
+
+open_locale big_operators
+open_locale classical
+open filter finset
+
+lemma card_le_div_nat {n p : ℕ} (hp : 0 < p) : card {e ∈ range n | p ∣ (e + 1)} ≤ n / p :=
+begin
+  let f : ℕ → ℕ := λ e, (e + 1) / p - 1,
+  let Np := {e ∈ range n | p ∣ (e + 1)},
+
+  have hf : ∀ a : ℕ, a ∈ Np → f a ∈ range (n / p),
+  { intros a ha,
+    simp only [Np, sep_def, mem_filter, mem_range] at ha,
+    obtain ⟨han, ⟨w, hw⟩⟩ := ha,
+    simp only [f, mem_range],
+    have hnp : n / p ≥ 1,
+    { rw ← nat.div_self hp,
+      exact nat.div_le_div_right (show n ≥ p, by nlinarith [(show w > 0, by nlinarith)]) },
+    calc (a + 1) / p - 1 ≤ n / p - 1 : nat.sub_le_sub_right
+                                         (nat.div_le_div_right (nat.succ_le_iff.mpr han)) 1
+    ...                  < n / p     : nat.sub_lt (by linarith) zero_lt_one },
+
+  have hf_inj : ∀ (a₁ : ℕ), a₁ ∈ Np → ∀ (a₂ : ℕ), a₂ ∈ Np → f a₁ = f a₂ → a₁ = a₂,
+  { intros a₁ ha₁ a₂ ha₂ hfeq,
+    simp only [f] at hfeq,
+    simp only [Np, sep_def, mem_filter, mem_range] at ha₁,
+    obtain ⟨hna₁, ⟨w₁, hw₁⟩⟩ := ha₁,
+    simp only [Np, sep_def, mem_filter, mem_range] at ha₂,
+    obtain ⟨hna₂, ⟨w₂, hw₂⟩⟩ := ha₂,
+    rw [hw₁, hw₂, nat.mul_div_cancel_left w₁ hp, nat.mul_div_cancel_left w₂ hp] at hfeq,
+    have hw₁_eq_w₂ : w₁ = w₂,
+    rw [← nat.succ_pred_eq_of_pos (show w₁ > 0, by nlinarith), ← nat.sub_one,
+        ← nat.succ_pred_eq_of_pos (show w₂ > 0, by nlinarith), ← nat.sub_one, hfeq],
+    rw [(show a₁ = p * w₁ - 1, by finish), (show a₂ = p * w₂ - 1, by finish), hw₁_eq_w₂] },
```

Comment:
These become so much simpler for you if you use lemmas about injectivity!
```suggestion
    have hw₁_eq_w₂ : w₁ = w₂, { refine nat.pred_inj _ _ hfeq; nlinarith },
    rw [← add_left_inj 1, hw₁, hw₂, hw₁_eq_w₂] },
```
Note that you could also use `nat.succ.inj_eq` instead of `add_left_inj`, but I chose the latter because you have everything in terms of `+ 1` and not `succ`.

Reply:
The `finish`es seemed to be making the proof extremely slow so I think this is a big improvement in that regard, too.